caiman/fiola/gpu_mc_nnls.py

```diff
@@ -0,0 +1,605 @@
+        else:
+            ncc = nominator    
+        
+        ncc = tf.where(tf.math.is_nan(ncc), tf.zeros_like(ncc), ncc)
+        sh_x, sh_y = self.extract_fractional_peak(ncc, self.ms_h, self.ms_w)
+        self.shifts = [sh_x, sh_y]
+        
+        #fr_corrected = tfa.image.translate(fr[0], (tf.squeeze(tf.stack([sh_y, sh_x], axis=1))), 
+        #                                    interpolation="bilinear")
+        fr_corrected = self.apply_shifts_dft_tf(fr[0], [-sh_x, -sh_y])
+        
+        if self.return_shifts:
+            return tf.reshape(tf.transpose(tf.squeeze(fr_corrected, axis=3), perm=[0,2,1]), (self.batch_size, self.shp_0[0]*self.shp_0[1])), self.shifts
+        else:
+            return tf.reshape(tf.transpose(tf.squeeze(fr_corrected, axis=3), perm=[0,2,1]), (self.batch_size, self.shp_0[0]*self.shp_0[1]))
+    
+    def normalize_template(self, template, epsilon=0.00000001):
+        # remove mean and divide by std
+        template_zm = template - tf.reduce_mean(template, axis=[0,1], keepdims=True)
+        template_var = tf.reduce_sum(tf.square(template_zm), axis=[0,1], keepdims=True) + epsilon
+        return template_zm, template_var
+        
+    def normalize_image(self, imgs, strides=[1,1,1,1], padding='VALID', epsilon=0.00000001):
+        # remove mean and standardize so that normalized cross correlation can be computed
+        if self.use_fft:
+            shape = [self.template.shape[0]-2*self.ms_w, self.template.shape[1]-2*self.ms_h]            
+        else:
+            shape = [self.template.shape[0], self.template.shape[1]]
+        imgs_zm = imgs - tf.reduce_mean(imgs, axis=[1,2], keepdims=True)
+        img_stack = tf.stack([imgs[:,:,:,0], tf.square(imgs)[:,:,:,0]], axis=3)
+        localsum_stack = tf.nn.avg_pool2d(img_stack,[1, shape[0], shape[1], 1], 
+                                               padding=padding, strides=strides)
+        localsum_ustack = tf.unstack(localsum_stack, axis=3)
+        localsum_sq = localsum_ustack[1][:,:,:,None]
+        localsum = localsum_ustack[0][:,:,:,None]      
+        imgs_var = localsum_sq - tf.square(localsum)/np.prod(shape) + epsilon
+        # Remove small machine precision errors after subtraction
+        imgs_var = tf.where(imgs_var<0, tf.zeros_like(imgs_var), imgs_var)
+        return imgs_zm, imgs_var        
+        
+    def extract_fractional_peak(self, ncc, ms_h, ms_w):
+        # use gaussian interpolation to extract a fractional shift
+        shifts_int = self.argmax_2d(ncc) 
+        shifts_int_cast = tf.cast(shifts_int,tf.int32)
+        sh_x, sh_y = shifts_int_cast[:,0],shifts_int_cast[:,1]
+        sh_x_n = tf.cast(-(sh_x - ms_h), tf.float32)
+        sh_y_n = tf.cast(-(sh_y - ms_w), tf.float32)
+        ncc_log = tf.math.log(ncc)
+
+        n_batches = np.arange(self.batch_size)
+        idx = tf.transpose(tf.stack([n_batches, tf.squeeze(sh_x-1,axis=1), tf.squeeze(sh_y,axis=1)]))
```

Comment:
Is there a more efficient, readable, and perhaps abstract way to write this block of code?